### ISSUE TITLE
Retain channel information from MIDI

### ISSUE THREAD

**Issue by nightscape** [cuthbertLab/music21#465]
We've got a somewhat related problem:
We're trying to read a MIDI file generated by a [Jammy guitar](https://playjammy.com/).
The Jammy uses 6 MIDI channels for the 6 strings of the guitar.
We would like to retain this information during MIDI => Music21 transformation in order to e.g. do analysis later on how a chord was fingered.
Is there some way to transform MIDI into a Music21 stream without losing the channel information?

Thanks!

(Note: This is a cross-post from here https://github.com/cuthbertLab/music21/issues/206, I wasn't sure if someone would see this in a closed ticket).

**Comment by mscuthbert** [comment]
I think that storing this information in the `.editorial` object would be a good place for it.

**Comment by jacobtylerwalls** [comment]
hiya @nightscape I put together a little proof of concept for this. Can you take it for a spin and let us know if it suits your needs? I was took your report to mean you had a single midi track for the guitar, and the notes in that track freely used 6 channels. Is that right?

**Comment by Meekohi** [comment]
Noticing this problem as well. If you a load a multi-track MIDI file, and then immediately save it, the tracks are assigned to seemingly arbitrary channels.

```
import copy
from music21 import *

orig = converter.parse('test/4part.mid')
orig.write('midi', fp='output.mid')
```
https://www.dropbox.com/s/oln4y19yn81kqhc/4part.mid?dl=0

The weirdest is that some of the tracks are assigned "Choir Ooos" which wasn't even in the original file. Must be some sort of logic currently in place based on track name?

**Comment by jacobtylerwalls** [comment]
> the tracks are assigned to seemingly arbitrary channels.

Yes, since channel information is not retained, it's created afresh. In your example I see channel 1 for the generic "soloist" and channel 2 for the SATB tracks, each of which have a 0-indexed midiProgram 53, see below...

See [channelInstrumentData()](http://web.mit.edu/music21/doc/moduleReference/moduleMidiTranslate.html?highlight=channelinstrumentdata#music21.midi.translate.channelInstrumentData).

> The weirdest is that some of the tracks are assigned "Choir Ooos" which wasn't even in the original file. Must be some sort of logic currently in place based on track name?

This should probably be tracked separately (program, rather than channel). What's happening is that the same routine is used to process both `PROGRAM_CHANGE` messages and `INSTRUMENT_NAME` or `SEQUENCE_TRACK_NAME` messages, and that routine attempts to set the midiProgram if the message has that data, which a ` SEQUENCE_TRACK_NAME` message will not. From the `SEQUENCE_TRACK_NAME` messages in your file, we normalized each of "Soprano", "Alto", etc. to an `instrument.Soprano`, etc., instance with default midiProgram = 53 (0-indexed, so = 54: Choir Oohs). We should be smarter about that.

**Comment by Meekohi** [comment]
I wonder if it would be practical to add some metadata to `stream.Part`, like `part.metadata.midiChannel = 1`, `part.metadata.midiInstrument = 1` or etc during loading, and then check for that during `.write` to return the MIDI back to the original condition (or someone could use that to micro-manage how they want the final midi output to look)?

**Comment by jacobtylerwalls** [comment]
For MIDI program changes, that should be possible today with inserting Instrument instances (which have a midiProgram attribute) into streams. (I believe we write patch change messages when we encounter instrument instances. Do let me know if I'm wrong about that.

The channel stuff is buried a bit deeper. But there would be a way to send an acceptableChannelList to channelInstrumentData() or even potentially override it.

I do think we should start capturing midi channel data (this ticket) and also avoiding false program data based on reading instrument name messages (the ticket I just created), so thanks for the report!

**Comment by mscuthbert** [comment]
Just a note that there won't be a guarantee ever that channel In = channel Out if pitch bends are involved, since we write different bends to different channels.

It's good to get many of these things in if easy (I don't want midiChannel on metadata; but it can go in editorial), but music21 is not and will never be a full MIDI<->MIDI platform, the same way we'd like to be with MusicXML, so some differences will always remain (and not be a bug)

**Comment by Meekohi** [comment]
Yeah totally understood @mscuthbert -- for us at least, if everything just defaulted to a "Piano" channel that would actually be fine, but the introduction of new channels that weren't in the original MIDI is causing us some issues. Hopefully there is some low-hanging fruit that will make things useful for simple MIDI applications that aren't using any complicated MIDI events or etc. 👍🏼

**Comment by duhaime** [comment]
I just wanted to mention that adding channel information will help users like [this developer on StackOverflow](https://stackoverflow.com/questions/41085082/extract-acoustic-bass-drum-from-midi-file-by-using-music21-library) who want to remove percussion from data files (as percussion is typically stored in channel 9 on midi). Should that channel be given special treatment to make sure e.g. pitch bends don't end up in there?

**Comment by thomasarvanitidis** [comment]
I'm a user like the one @duhaime mentioned above and such fix would make music21 a viable package for the drums MIDI generation community. I also noticed that a "Pitch Bend" event is also added when simply loading -> writing a drums MIDI file with music21, which I don't understand why it happens.
